### ISSUE TITLE
OverlayFS creates duplicates when moving with Vifm

### ISSUE THREAD

**Issue by aleksejrs** [vifm/vifm#375]
There is OverlayFS in the Linux kernel, see `https://en.wikipedia.org/wiki/OverlayFS` and `/usr/share/doc/linux-doc/Documentation/filesystems/overlayfs.txt.gz` on Debian.

I haven't tested this intentionally, but I can see its effects: if I move files/directories inside a mount (not "Changes to underlying filesystems", but often between (sub)subdirectories of sibling directories on different underlying filesystems) using Vifm, it's slow and leaves duplicate files.  The mounts are mostly on btrfs.  `mv` has no issues: it works instantly and so uses no extra space.

**Comment by xaizek** [comment]
> I haven't tested this intentionally,

I tried and my results differ, but maybe you meant something different. Here is my setup:
```bash
mkdir -p /tmp/overlay/{lower,upper,mount,workdir} /tmp/overlay/lower/low /tmp/overlay/upper/up
echo low > /tmp/overlay/lower/low/in-lower
echo up > /tmp/overlay/upper/up/in-upper
echo low > /tmp/overlay/lower/in-both
echo up > /tmp/overlay/upper/in-both

mount -t overlay overlay -olowerdir=/tmp/overlay/lower,upperdir=/tmp/overlay/upper,workdir=/tmp/overlay/workdir /tmp/overlay/mount
```
Moving `/tmp/overlay/upper/up/in-upper` works fine, but moving `/tmp/overlay/lower/low/in-lower` or `/tmp/overlay/upper/in-both` doesn't work with either Vifm ("rename operation failed") or `mv` ("won't move directory under itself").

Do you get different results in that setup?

**Comment by aleksejrs** [comment]
TLDR: The problem exists with `mv`. Closing the report.

I get the same results as you do with your setup. I meant something different; I am not sure what exactly to do, but you need multiple lower directories, probably with subdirectories in them, because maybe it's caused when I rename the subdirectories until their names clash. There were special files left in the directories at first, but finally there was a duplicate file in one.

Try this script:

```shell
mkdir -p /tmp/overlay/{upper,mount,workdir}
mkdir -p /tmp/overlay/lower{1,2}
mkdir -p /tmp/overlay/lower1/dir-in-lower1
mkdir -p /tmp/overlay/lower2/dir-in-lower2
touch /tmp/overlay/lower1/dir-in-lower1/file-from-dir-in-lower1
touch /tmp/overlay/lower2/dir-in-lower2/file-from-dir-in-lower2
mount -t overlay overlay -olowerdir=/tmp/overlay/lower1:/tmp/overlay/lower2,upperdir=/tmp/overlay/upper,workdir=/tmp/overlay/workdir /tmp/overlay/mount
```

Use mv or Vifm to move `mount/dir-in-lower1/file-from-dir-in-lower1` to `mount/dir-in-lower2/`

What I get:

`mount/` contains:
* `dir-in-lower1` (empty)
* `dir-in-lower2` with `file-from-dir-in-lower1` and `file-from-dir-in-lower2`

* `lower1/` contains `dir-in-lower1` with `file-from-dir-in-lower1`
* `lower2/` contains `dir-in-lower2` with `file-from-dir-in-lower2`

upper contains
* `dir-in-lower1` with with a special file `file-from-dir-in-lower1`.
* `dir-in-lower2` with with `file-from-dir-in-lower2`.

Maybe OverlayFS principles or options (or bugs) do not allow it to modify lower directories.

**Comment by aleksejrs** [comment]
Apparently it's not even supposed to work at all: https://serverfault.com/questions/765441/overlayfs-how-can-i-delete-files-not-hide-them
